src/rgw/h3/src/connection.cc

```diff
@@ -0,0 +1,555 @@
+      ex(asio::make_strand(socket.get_executor())),
+      socket(socket), on_new_stream(on_new_stream),
+      timeout_timer(ex), pacing_timer(ex),
+      conn(std::move(_conn)), cid(std::move(cid))
+{
+}
+
+ConnectionImpl::~ConnectionImpl()
+{
+  observer.on_conn_destroy(cid);
+}
+
+auto ConnectionImpl::accept()
+    -> asio::awaitable<error_code, executor_type>
+{
+  observer.on_conn_accept(cid);
+
+  reset_timeout(); // schedule the initial timeout
+
+  co_return co_await writer();
+}
+
+void ConnectionImpl::reset_timeout()
+{
+  auto ns = std::chrono::nanoseconds{::quiche_conn_timeout_as_nanos(conn.get())};
+  observer.on_conn_schedule_timeout(cid, ns);
+  timeout_timer.expires_after(ns);
+  if (ns.count()) {
+    timeout_timer.async_wait([this] (error_code ec) { on_timeout(ec); });
+  }
+}
+
+void ConnectionImpl::on_timeout(error_code ec)
+{
+  if (ec) {
+    return;
+  }
+
+  ::quiche_conn_on_timeout(conn.get());
+
+  if (::quiche_conn_is_closed(conn.get())) {
+    ec = on_closed();
+  }
+
+  writer_wake(ec);
+}
+
+auto ConnectionImpl::flush_some()
+    -> asio::awaitable<error_code, executor_type>
+{
+  // send up to 8 packets at a time with sendmmsg()
+  static constexpr size_t max_mmsg = 8;
+  std::array<message, max_mmsg> messages;
+  std::array<iovec, max_mmsg> iovs;
+  std::array<mmsghdr, max_mmsg> headers;
+  std::array<quiche_send_info, max_mmsg> sendinfos;
+
+  error_code ec;
+  size_t count = 0;
+  for (; count < max_mmsg; count++) {
+    auto& m = messages[count];
+    auto& sendinfo = sendinfos[count];
+
+    // serialize the next packet for sending
+    const ssize_t bytes = ::quiche_conn_send(
+        conn.get(), m.buffer.data(), m.buffer.max_size(), &sendinfo);
+
+    if (bytes < 0) {
+      ec.assign(bytes, quic_category());
+      if (ec == quic_errc::done) {
+        if (count == 0) {
+          co_return ec;
+        }
+        break; // send the packets we've already prepared
+      }
+      observer.on_conn_send_error(cid, ec);
+      co_return ec;
+    }
+
+    auto& iov = iovs[count];
+    iov = iovec{m.buffer.data(), static_cast<size_t>(bytes)};
+
+    auto& h = headers[count];
+    h.msg_hdr.msg_name = &sendinfo.to;
+    h.msg_hdr.msg_namelen = sendinfo.to_len;
+    h.msg_hdr.msg_iov = &iov;
+    h.msg_hdr.msg_iovlen = 1;
+    h.msg_hdr.msg_control = nullptr;
+    h.msg_hdr.msg_controllen = 0;
+    h.msg_hdr.msg_flags = 0;
+    h.msg_len = 0;
+  }
+
+  // quiche's pacing hints have undefined behavior and can return tv_nsec
+  // outside the specified range, leading to huge pacing delays:
+  // https://github.com/cloudflare/quiche/pull/1403
+#if 0
+  // apply the pacing delay from the last packet
+  const auto& last = sendinfos[count-1];
+  const auto send_at = pacing_clock::time_point{
+      std::chrono::seconds(last.at.tv_sec) +
+      std::chrono::nanoseconds(last.at.tv_nsec)};
+
+  if (send_at > pacing_clock::zero()) {
+    const auto now = pacing_clock::now();
+    if (send_at > now) {
```

Comment:
in testing, the condition code flow wasn't executing - `send_at` was never larger than `now`
```suggestion
    if (now > send_at) {
```